### PR TITLE
System Tests for Mlrun Demos

### PR DESCRIPTION
**Pull request by quaark**
Using the demos from https://github.com/mlrun/demos.
Added system tests that use the system as expected with the horovod, churn and sklearn demos and ensure the pipelines work correctly.

### CODE REVIEW

File: setup.cfg [hunk 1]
```diff
@@ -3,4 +3,4 @@ test = pytest
 
 [tool:pytest]
 addopts = -v -rf --disable-warnings
-python_files = tests/*.py
+python_files = tests/*/test_*.py
```
Comment (Hedingber) [inline setup.cfg:6]: I see that this can be a list (https://docs.pytest.org/en/stable/reference.html#confval-python_files), e.g.
`python_files = tests/*.py tests/*/test_*.py`
are you sure the new path include everything that the old path included ? why this change is needed

File: tests/system/demos/base.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+    project_name = ''
+
+    def custom_setup(self):
+        mlrun.set_environment(
+            artifact_path='/User/data', project=self.project_name,
+        )
+        self._artifact_path = os.path.join(
+            mlrun.mlconf.artifact_path, 'pipeline/{{workflow.uid}}'
+        )
```
Comment (Hedingber) [inline tests/system/demos/base.py:18]: Why `artifact_path` in the environment and in `self` are different ?

File: tests/system/demos/base.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+        self._demo_project = self.create_demo_project()
+
+        self._logger.debug(
+            'Project Ready',
```
Comment (Hedingber) [inline tests/system/demos/base.py:22]:
```suggestion
            'Created demo project',
```
?

File: tests/system/demos/base.py [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+        db.del_artifacts(tag='*', project=self.project_name)
+
+    def create_demo_project(self) -> mlrun.projects.MlrunProject:
+        raise NotImplementedError
+
+    def run_and_verify_project(self, runs_amount: int = 1, arguments: dict = None):
+        arguments = arguments or {}
+        run_id = self._demo_project.run(
+            'main',
+            arguments=arguments,
```
Comment (Hedingber) [inline tests/system/demos/base.py:40]: Are you sure there is an `arguments` kwarg ?

File: tests/system/demos/base.py [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+            artifact_path=self._artifact_path,
+            dirty=True,
+            watch=True,
+        )
+
+        db = mlrun.get_run_db().connect()
```
Comment (Hedingber) [inline tests/system/demos/base.py:46]: Maybe will be nicer to initialize it once and then use as `self.run_db` ?

File: tests/system/demos/horovod/test_horovod.py [hunk 6]
```diff
@@ -0,0 +1,78 @@
+import pathlib
+
+import mlrun
+
+from tests.system.base import TestMLRunSystem
+from tests.system.demos.base import TestDemo
+
+
+@TestMLRunSystem.skip_test_if_env_not_configured
+class TestHorovodTSv2(TestDemo):
```
Comment (Hedingber) [inline tests/system/demos/horovod/test_horovod.py:10]:
```suggestion
class TestHorovodTFv2(TestDemo):
```

File: tests/system/demos/horovod/test_horovod.py [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+        mlrun.mount_v3io()
+
+        self._logger.debug('Uploading training file')
+        trainer_src_path = str(self.assets_path / 'horovod_training.py')
+        trainer_dest_path = pathlib.Path('/assets/horovod_training.py')
+        stores = mlrun.datastore.store_manager.set()
+        datastore, subpath = stores.get_or_create_store(
+            self._get_v3io_user_store_path(trainer_dest_path)
+        )
+        datastore.upload(subpath, trainer_src_path)
```
Comment (Hedingber) [inline tests/system/demos/horovod/test_horovod.py:29]: That part looks like 💩  - not your fault of course, you're just using the existing SDK functions,
I honestly don't understand what each line is doing here, if you have a vision for how to make the data stores SDK more clear - trello it

File: tests/system/demos/churn/test_churn.py [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+
+        self._logger.debug('Setting project workflow')
+        demo_project.set_workflow(
+            'main', str(self.assets_path / 'workflow.py'), embed=True
+        )
+
+        return demo_project
+
+    def test_demo(self):
+        self.run_and_verify_project(runs_amount=6)
```
Comment (Hedingber) [inline tests/system/demos/churn/test_churn.py:56]: This is a magic number that I'm not sure where it's coming from, I would guess it should equal the number of functions in the project - but I see 7 not 6 (maybe there are functions we set but don't use ?)
Ideally let's base this on some variable (like `len(demo_project.functions)`) so when it will change the test will work
If that's not easy let's at least add a comment explaining where this number coming from

File: tests/system/demos/horovod/test_horovod.py [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+        self._logger.debug('Setting project workflow')
+        demo_project.set_workflow(
+            'main', str(self.assets_path / 'workflow.py'), embed=True
+        )
+
+        return demo_project
+
+    def test_demo(self):
+        self.run_and_verify_project(
+            runs_amount=3,
```
Comment (Hedingber) [inline tests/system/demos/horovod/test_horovod.py:73]: same comment as for previous demo `run_amount`

File: tests/system/demos/sklearn/test_sklearn.py [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+
+        self._logger.debug('Setting project workflow')
+        demo_project.set_workflow(
+            'main', str(self.assets_path / 'workflow.py'), embed=True
+        )
+
+        return demo_project
+
+    def test_demo(self):
+        self.run_and_verify_project(runs_amount=5)
```
Comment (Hedingber) [inline tests/system/demos/sklearn/test_sklearn.py:48]: same

**Review by Hedingber — CHANGES_REQUESTED** [review]
Looks awesome!

File: setup.cfg (hunk 1 above)
Comment (quaark) [reply setup.cfg:6]: Because this imported the assets in to pytest, and so the training files that import `tensorflow` an `horovod` failed cause the packages aren't installed.
And yes I checked, we didn't loose coverage. Everything is by convention with the `test_` prefix.

File: tests/system/demos/base.py (hunk 4 above)
Comment (quaark) [reply tests/system/demos/base.py:41]: yes, I use it in the `horovod` demo

File: tests/system/demos/horovod/test_horovod.py (hunk 7 above)
Comment (quaark) [reply tests/system/demos/horovod/test_horovod.py:29]: Yeah I know, that is exactly why it took me over a day to figure out how to use it.
I'll think of something

File: tests/system/demos/base.py (hunk 2 above)
Comment (quaark) [reply tests/system/demos/base.py]: The first is the general one, and the second is run specific, I'll change the name

File: tests/system/demos/churn/test_churn.py (hunk 8 above)
Comment (quaark) [reply tests/system/demos/churn/test_churn.py:56]: The seventh you see is the `deploy_step` at the end, the other 6 are the `as_step` in the workflow pipeline.
Not sure if I can make this not magic...

File: tests/system/demos/horovod/test_horovod.py (hunk 9 above)
Comment (quaark) [reply tests/system/demos/horovod/test_horovod.py:76]: same

File: tests/system/demos/sklearn/test_sklearn.py (hunk 10 above)
Comment (quaark) [reply tests/system/demos/sklearn/test_sklearn.py:51]: same

**Review by Hedingber — APPROVED** [review]
